shell/store/notifications.ts

```diff
@@ -41,9 +44,58 @@ export const state = function(): NotificationsStore {
   };
 };
 
-const debounceSetNotifications = debounce((state: NotificationsStore, notifications: StoredNotification[]) => {
-  state.notifications = notifications;
-}, 500);
+let bc: BroadcastChannel;
+
+function sync(operation: string, param?: any) {
+  bc?.postMessage({ operation, param });
+}
+
+async function saveEncryptedNotification(getters: any, notification: Notification) {
+  const toEncrypt: EncryptedNotification = {
+    title:           notification.title,
+    message:         notification.message,
+    level:           notification.level,
+    primaryAction:   notification.primaryAction,
+    secondaryAction: notification.secondaryAction,
+  };
+
+  const localStorageKey = getters['localStorageKey'];
+  const encryptionKey = getters['encryptionKey'];
+
+  try {
+    const data = JSON.stringify(toEncrypt);
+    const enc = await encrypt(data, encryptionKey);
+
+    window.localStorage.setItem(`${ localStorageKey }-${ notification.id }`, JSON.stringify(enc));
+  } catch (e) {
+    console.error('Unable to save notification to local storage', e); // eslint-disable-line no-console
+  }
+}
+
+/**
+ * Sync the notifications index to local storage.
+ *
+ * We only store the non-sensitive data about the notifications in the index - the other data is stored in individual entries which are encrypted.
+ */
+function syncIndex(state: NotificationsStore) {
```

Comment:
this (or somewhere) should also remove (current user) notifications that aren't in (current user's) state.notifications (they've been `pop`ed on `add`).